### PR TITLE
feat(mobile): lazy loading of assets

### LINKED ISSUE #1547: [Tech-Debt]: (mobile) Unify asset grid of albums, shared albums and favorites
### Feature detail

Unify asset grid of albums, shared albums and favorites

### Platform

Mobile App

### LINKED ISSUE #2287: Archived asset cannot be favorite
### The bug

I will deem this a bug as archived asset gets favorite tag removed and it cannot be set.

Screenshot_20230419-092044

### The OS that Immich Server is running on

Kubernetes

### Version of Immich Server

1.54.0

### Version of Immich Mobile App

1.54.0

### Platform with the issue

- [X] Server
- [ ] Web
- [X] Mobile

### Your docker-compose.yml content

```YAML
Not relevant
```

### Your .env content

```Shell
Not relevant
```

### Reproduction steps

```bash
1. Find an asset and set it as favorite
2. Check that it can be found in favorites view
3. Archive it
4. Notice that you cannot find it in favorites anymore and asset debug info show that it's not favourite anymore.
```

### Additional information

The screenshot is from when I starred it without leaving the viewer but it's not saved to the asset.

### PR DESCRIPTION
**Pull request by fyfrey**
reworks how asset metadata is loaded in the app

allows to display huge asset collections on timeline and in albums

- [x] all screens now use the asset grid
- [x] no longer loads all assets to render the list. instead batches are fetched from the local database as needed
- [x] simplified state management of assets: all used providers are directly updated when the Isar database changes
- [x] removes 500+ LOC :-)
- [x] adds a new asset grouping mode "automatic": if there are few assets in a month, groups by month. If there are many, groups by day within that month. If there are too many assets on a single day, split into time ranges
- [x] album header now shrinks in size when asset grid is scrolled
- [x] favorite provider is no longer needed to render assets: assets have `isFavorite`
- [x] asset detail viewer watches asset for changes (e.g. sync running in background and isFavorite is changed OR it is merged with remote info etc.)

### CODE REVIEW

**Comment by matthinc** [comment]
Screenshot 2023-05-09 at 19 10 06

Little display bug while drag-scrolling

File: mobile/lib/modules/favorite/providers/favorite_provider.dart [hunk 1]
```diff
@@ -1,68 +1,77 @@
...
+  yield await RenderList.fromQuery(
+    query,
+    GroupAssetsBy.values[settings.getSetting(AppSettingsEnum.groupAssetsBy)],
+  );
+  await for (final _ in query.watchLazy()) {
+    yield await RenderList.fromQuery(
+      query,
+      GroupAssetsBy.values[settings.getSetting(AppSettingsEnum.groupAssetsBy)],
+    );
+  }
```
Comment (matthinc) [inline mobile/lib/modules/favorite/providers/favorite_provider.dart:76]: Cool!

File: mobile/lib/modules/home/ui/asset_grid/asset_grid_data_structure.dart [hunk 2]
```diff
@@ -2,212 +2,282 @@ import 'dart:math';
...
+        _buf = query!.offset(start).limit(len).findAllSync();
+        _bufOffset = start;
+      }
+      assert(_bufOffset <= offset);
+      assert(_bufOffset + _buf.length >= offset + count);
+      final arr = _buf.slice(offset - _bufOffset, offset - _bufOffset + count);
+      return arr;
     }
+    throw Exception("RenderList has neither assets nor query");
+  }
```
Comment (matthinc) [inline mobile/lib/modules/home/ui/asset_grid/asset_grid_data_structure.dart:84]: Can you add some comments explaining your algorithm, please? This code is (at least for me) very complex and hard to understand.

**Comment by matthinc** [comment]
I really like the approach of lazy-loading assets only when needed. We probably have to come up with another drag-scrolling mechanism since the current one relies on data from all assets to build its layout properly. Do you have any idea in mind for alternatives?

**Comment by fyfrey** [comment]
> Screenshot 2023-05-09 at 19 10 06
> 
> Little display bug while drag-scrolling

What's the bug? showing simple bars instead of laying out all assets is a necessary simplification as we don't know the assets while scrolling :-)

**Comment by matthinc** [comment]
> > Screenshot 2023-05-09 at 19 10 06
> > Little display bug while drag-scrolling
> 
> What's the bug? showing simple bars instead of laying out all assets is a simplification as we don't know the assets while scrolling :-) we could of course always make square boxes according to the number of assets per row setting.. I'll look into that. If there's another issue please notify me

Ok, those boxes looked buggy to me as I knew the previous behavior. I don't think there is any point in showing them at all if they don't represent anything. What about just leaving the screen black then?

**Comment by fyfrey** [comment]
> Ok, those boxes looked buggy to me as I knew the previous behavior. I don't think there is any point in showing them at all if they don't represent anything. What about just leaving the screen black then?

We need something to fill the exact height so the scrolling behaves correctly. Showing nothing (black screen) will probably be weird while scrolling a long time without headlines from days/months. I'll rework this to be similar to the previous behavior.

File: mobile/lib/modules/home/ui/asset_grid/asset_grid_data_structure.dart (hunk 2 above)
Comment (fyfrey) [reply mobile/lib/modules/home/ui/asset_grid/asset_grid_data_structure.dart]: sure, added comments

File: mobile/lib/modules/album/views/album_viewer_page.dart [hunk 3]
```diff
@@ -241,75 +200,71 @@ class AlbumViewerPage extends HookConsumerWidget {
...
-              ),
+        data: (data) => WillPopScope(
+          onWillPop: onWillPop,
+          child: GestureDetector(
+            onTap: () {
+              titleFocusNode.unfocus();
+            },
+            child: Column(
+              children: [
+                // TODO header takes up space while scrolling through the assets
```
Comment (fyfrey) [inline mobile/lib/modules/album/views/album_viewer_page.dart:244]: @martyfuhry can you take a look at this issue if you have some time? thanks!

File: mobile/lib/modules/home/ui/asset_grid/immich_asset_grid_view.dart [hunk 4]
```diff
@@ -241,14 +333,33 @@ class ImmichAssetGridViewState extends State<ImmichAssetGridView> {
...
     super.dispose();
   }
 
+  void _positionListener() {
+    final values = _itemPositionsListener.itemPositions.value;
+    final start = values.firstOrNull;
+    final end = values.lastOrNull;
+    if (start != null && end != null) {
+      if (start.index <= end.index) {
+        widget.visibleItemsListener!(start, end);
```
Comment (martyfuhry) [inline mobile/lib/modules/home/ui/asset_grid/immich_asset_grid_view.dart:356]: `widget.visibleItemsListener?.call(start, end)` is safer. Even though this is properly guarded at line 344, the optional `.?call` will be more maintainable in the future.

File: mobile/lib/modules/home/ui/asset_grid/immich_asset_grid_view.dart [hunk 5]
```diff
@@ -241,14 +333,33 @@ class ImmichAssetGridViewState extends State<ImmichAssetGridView> {
...
+      } else {
+        widget.visibleItemsListener!(end, start);
```
Comment (martyfuhry) [inline mobile/lib/modules/home/ui/asset_grid/immich_asset_grid_view.dart:358]: Same here.

**Comment by alextran1502** [comment]
Hello Fynn, I've done my bit of testing, and below are a few things I found.

1. During the database indexing, and upserting process upon a new login, we don't have any indication. This might confuse people.
2. When switching users by logging out and logging in the newly logged-in user has the database indexing completed but doesn't have the timeline filled in, as well as the library page. The sharing page correctly displays the albums. In order to get the timeline to show the asset and the library page to show the albums I have to restart the app. This might be the problem with the stream listener?

Let me know if you need anything else

**Comment by fyfrey** [comment]
> 1. During the database indexing, and upserting process upon a new login, we don't have any indication. This might confuse people.

Indeed. I've fixed this by showing the loading indicator instead if the grid would be empty 

> 2. When switching users by logging out and logging in the newly logged-in user has the database indexing completed but doesn't have the timeline filled in, as well as the library page. The sharing page correctly displays the albums. In order to get the timeline to show the asset and the library page to show the albums I have to restart the app. This might be the problem with the stream listener?

This was tricky. The providers are not normally reset while the app stays open... And the user id is thus cached in it and stays the same after changing logged in user. I've fixed this by make the providers auto dispose, so that they are destroyed when the view they're used on is destroyed.

File: mobile/lib/shared/providers/asset.provider.dart [hunk 6]
```diff
@@ -220,80 +120,79 @@ class AssetNotifier extends StateNotifier<AssetsState> {
...
+      .ownerIdEqualTo(Store.get(StoreKey.currentUser).isarId)
+      .isArchivedEqualTo(false)
+      .sortByFileCreatedAtDesc();
+  final settings = ref.watch(appSettingsServiceProvider);
+  final groupBy =
+      GroupAssetsBy.values[settings.getSetting(AppSettingsEnum.groupAssetsBy)];
+  yield await RenderList.fromQuery(query, groupBy);
+  await for (final _ in query.watchLazy()) {
+    yield await RenderList.fromQuery(query, groupBy);
+  }
```
Comment (alextran1502) [inline mobile/lib/shared/providers/asset.provider.dart:178]: I saw this implementation across the PR. Am I understanding correctly that on line 175 we return the result the first time this provider is watched/called? Then we setup the a listener to the subsequent changes on line 176-178 for any future changes?

**Comment by alextran1502** [comment]
I found a bug that when you click on favorite button for the first time, it showed your an error below

```
I/flutter (18024): [SEVERE] [2023-05-16 10:53:23.546425] 'package:immich_mobile/shared/models/asset.dart': Failed assertion: line 236 pos 12: 'canUpdate(a)': is not true.
```

Then, clicking on it again would work.

**Review by alextran1502 — APPROVED** [review]
Thank you for another high-quality PR. I only have a few more questions, and this should be good to merge!

File: mobile/lib/modules/album/providers/shared_album.provider.dart [hunk 7]
```diff
@@ -75,21 +56,32 @@ class SharedAlbumNotifier extends StateNotifier<List<Album>> {
...
-    FutureProvider.autoDispose.family<Album?, int>((ref, albumId) async {
+    StreamProvider.autoDispose.family<Album, int>((ref, albumId) async* {
   final AlbumService sharedAlbumService = ref.watch(albumServiceProvider);
 
-  final Album? a = await sharedAlbumService.getAlbumDetail(albumId);
-  await a?.loadSortedAssets();
-  return a;
+  await for (final a in sharedAlbumService.watchAlbum(albumId)) {
+    if (a == null) {
+      throw Exception("Album with ID=$albumId does not exist anymore!");
```
[the remark this answers is not in the document]
Comment (fyfrey) [reply mobile/lib/modules/album/providers/shared_album.provider.dart:81]: No, it prints the error on the screen and the user can navigate back to the previous page.

File: mobile/lib/shared/providers/asset.provider.dart (hunk 6 above)
Comment (fyfrey) [reply mobile/lib/shared/providers/asset.provider.dart:178]: Yeah, that is correct!

**Comment by fyfrey** [comment]
> Thank you for another high-quality PR. I only have a few more questions, and this should be good to merge!

Added minor improvements for the two points you raised. This should be good to merge now.

**Comment by alextran1502** [comment]
@fyfrey I tested the changes you made, and it appears to cause some issues with rerendering/updating the page when an asset is archived/unarchived; Also, when clicking on the favorite button, the change isn't updated even though there is a message of `upserted x assets into the DB`, you will have to tap on the icon again for it to execute.

**Pull request merged**